File: archive/pulumi-py/Pr6148/lib/vpc_component.py
import pulumi
import pulumi_aws as aws
from pulumi import ComponentResource, ResourceOptions


class VpcComponent(ComponentResource):
    """
    Reusable VPC component with public and private subnets across multiple AZs
    """

    def __init__(
        self,
        name: str,
        environment_suffix: str,
        cidr_block: str,
        availability_zones: list,
        tags: dict,
        opts: ResourceOptions = None,
    ):
        super().__init__("custom:network:VpcComponent", name, None, opts)

        # Create VPC
        self.vpc = aws.ec2.Vpc(
            f"vpc-{environment_suffix}",
            cidr_block=cidr_block,
            enable_dns_hostnames=True,
            enable_dns_support=True,
            tags={**tags, "Name": f"vpc-{environment_suffix}"},
            opts=ResourceOptions(parent=self),
        )

        # Create Internet Gateway
        self.igw = aws.ec2.InternetGateway(
            f"igw-{environment_suffix}",
            vpc_id=self.vpc.id,
            tags={**tags, "Name": f"igw-{environment_suffix}"},
            opts=ResourceOptions(parent=self),
        )

        # Create public subnets
        self.public_subnets = []
        for i, az in enumerate(availability_zones):
            subnet = aws.ec2.Subnet(
                f"public-subnet-{i}-{environment_suffix}",
                vpc_id=self.vpc.id,
                cidr_block=f"10.0.{i}.0/24",
                availability_zone=az,
                map_public_ip_on_launch=True,
                tags={**tags, "Name": f"public-subnet-{i}-{environment_suffix}"},
                opts=ResourceOptions(parent=self),
            )
            self.public_subnets.append(subnet)

        # Create private subnets
        self.private_subnets = []
        for i, az in enumerate(availability_zones):
            subnet = aws.ec2.Subnet(
                f"private-subnet-{i}-{environment_suffix}",
                vpc_id=self.vpc.id,
                cidr_block=f"10.0.{i+10}.0/24",
                availability_zone=az,
                tags={**tags, "Name": f"private-subnet-{i}-{environment_suffix}"},
                opts=ResourceOptions(parent=self),
            )
            self.private_subnets.append(subnet)

        # Create NAT Gateway (single for cost optimization)
        self.eip = aws.ec2.Eip(
            f"nat-eip-{environment_suffix}",
            domain="vpc",
            tags={**tags, "Name": f"nat-eip-{environment_suffix}"},
            opts=ResourceOptions(parent=self),
        )

        self.nat_gateway = aws.ec2.NatGateway(
            f"nat-{environment_suffix}",
            subnet_id=self.public_subnets[0].id,
            allocation_id=self.eip.id,
            tags={**tags, "Name": f"nat-{environment_suffix}"},
            opts=ResourceOptions(parent=self),
        )

        # Create public route table
        self.public_route_table = aws.ec2.RouteTable(
            f"public-rt-{environment_suffix}",
            vpc_id=self.vpc.id,
            tags={**tags, "Name": f"public-rt-{environment_suffix}"},
            opts=ResourceOptions(parent=self),
        )

        # Public route to Internet Gateway
        aws.ec2.Route(
            f"public-route-{environment_suffix}",
            route_table_id=self.public_route_table.id,
            destination_cidr_block="0.0.0.0/0",
            gateway_id=self.igw.id,
            opts=ResourceOptions(parent=self),
        )

        # Associate public subnets with public route table
        for i, subnet in enumerate(self.public_subnets):
            aws.ec2.RouteTableAssociation(
                f"public-rta-{i}-{environment_suffix}",
                subnet_id=subnet.id,
                route_table_id=self.public_route_table.id,
                opts=ResourceOptions(parent=self),
            )

        # Create private route table
        self.private_route_table = aws.ec2.RouteTable(
            f"private-rt-{environment_suffix}",
            vpc_id=self.vpc.id,
            tags={**tags, "Name": f"private-rt-{environment_suffix}"},
            opts=ResourceOptions(parent=self),
        )

        # Private route to NAT Gateway
        aws.ec2.Route(
            f"private-route-{environment_suffix}",
            route_table_id=self.private_route_table.id,
            destination_cidr_block="0.0.0.0/0",
            nat_gateway_id=self.nat_gateway.id,
            opts=ResourceOptions(parent=self),
        )

        # Associate private subnets with private route table
        for i, subnet in enumerate(self.private_subnets):
            aws.ec2.RouteTableAssociation(
                f"private-rta-{i}-{environment_suffix}",
                subnet_id=subnet.id,
                route_table_id=self.private_route_table.id,
                opts=ResourceOptions(parent=self),
            )

        # Export properties
        self.vpc_id = self.vpc.id
        self.public_subnet_ids = [s.id for s in self.public_subnets]
        self.private_subnet_ids = [s.id for s in self.private_subnets]

        self.register_outputs(
            {
                "vpc_id": self.vpc_id,
                "public_subnet_ids": self.public_subnet_ids,
                "private_subnet_ids": self.private_subnet_ids,
            }
        )

File: archive/pulumi-ts/Pr6025/lib/vpc-component.ts
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface VpcComponentArgs {
  environmentName: string;
  vpcCidr: string;
  availabilityZones: string[];
  environmentSuffix: string;
  tags?: { [key: string]: string };
}

export class VpcComponent extends pulumi.ComponentResource {
  public readonly vpc: aws.ec2.Vpc;
  public readonly publicSubnets: aws.ec2.Subnet[];
  public readonly privateSubnets: aws.ec2.Subnet[];
  public readonly internetGateway: aws.ec2.InternetGateway;
  public readonly natGateways: aws.ec2.NatGateway[];
  public readonly webSecurityGroup: aws.ec2.SecurityGroup;
  public readonly appSecurityGroup: aws.ec2.SecurityGroup;
  public readonly flowLogGroup: aws.cloudwatch.LogGroup;

  constructor(
    name: string,
    args: VpcComponentArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('custom:network:VpcComponent', name, {}, opts);

    const defaultTags = {
      Environment: args.environmentName,
      ManagedBy: 'Pulumi',
      CostCenter: 'Platform',
      ...args.tags,
    };

    // Create VPC
    this.vpc = new aws.ec2.Vpc(
      `vpc-${args.environmentName}-${args.environmentSuffix}`,
      {
        cidrBlock: args.vpcCidr,
        enableDnsHostnames: true,
        enableDnsSupport: true,
        tags: {
          Name: `${args.environmentName}-vpc-${args.environmentSuffix}`,
          ...defaultTags,
        },
      },
      { parent: this }
    );

    // Create Internet Gateway
    this.internetGateway = new aws.ec2.InternetGateway(
      `igw-${args.environmentName}-${args.environmentSuffix}`,
      {
        vpcId: this.vpc.id,
        tags: {
          Name: `${args.environmentName}-igw-${args.environmentSuffix}`,
          ...defaultTags,
        },
      },
      { parent: this }
    );

    // Create Public Subnets
    this.publicSubnets = [];
    args.availabilityZones.forEach((az, index) => {
      const subnet = new aws.ec2.Subnet(
        `public-subnet-${args.environmentName}-${az}-${args.environmentSuffix}`,
        {
          vpcId: this.vpc.id,
          cidrBlock: `${args.vpcCidr.split('/')[0].split('.').slice(0, 2).join('.')}.${index * 2}.0/24`,
          availabilityZone: az,
          mapPublicIpOnLaunch: true,
          tags: {
            Name: `${args.environmentName}-public-${az}-${args.environmentSuffix}`,
            Type: 'Public',
            ...defaultTags,
          },
        },
        { parent: this }
      );
      this.publicSubnets.push(subnet);
    });

    // Create Private Subnets
    this.privateSubnets = [];
    args.availabilityZones.forEach((az, index) => {
      const subnet = new aws.ec2.Subnet(
        `private-subnet-${args.environmentName}-${az}-${args.environmentSuffix}`,
        {
          vpcId: this.vpc.id,
          cidrBlock: `${args.vpcCidr.split('/')[0].split('.').slice(0, 2).join('.')}.${index * 2 + 1}.0/24`,
          availabilityZone: az,
          tags: {
            Name: `${args.environmentName}-private-${az}-${args.environmentSuffix}`,
            Type: 'Private',
            ...defaultTags,
          },
        },
        { parent: this }
      );
      this.privateSubnets.push(subnet);
    });

    // Create Elastic IPs for NAT Gateways
    const eips: aws.ec2.Eip[] = [];
    args.availabilityZones.forEach((az, _index) => {
      const eip = new aws.ec2.Eip(
        `nat-eip-${args.environmentName}-${az}-${args.environmentSuffix}`,
        {
          domain: 'vpc',
          tags: {
            Name: `${args.environmentName}-nat-eip-${az}-${args.environmentSuffix}`,
            ...defaultTags,
          },
        },
        { parent: this }
      );
      eips.push(eip);
    });

    // Create NAT Gateways
    this.natGateways = [];
    args.availabilityZones.forEach((az, index) => {
      const nat = new aws.ec2.NatGateway(
        `nat-${args.environmentName}-${az}-${args.environmentSuffix}`,
        {
          subnetId: this.publicSubnets[index].id,
          allocationId: eips[index].id,
          tags: {
            Name: `${args.environmentName}-nat-${az}-${args.environmentSuffix}`,
            ...defaultTags,
          },
        },
        { parent: this }
      );
      this.natGateways.push(nat);
    });

    // Create Public Route Table
    const publicRouteTable = new aws.ec2.RouteTable(
      `public-rt-${args.environmentName}-${args.environmentSuffix}`,
      {
        vpcId: this.vpc.id,
        tags: {
          Name: `${args.environmentName}-public-rt-${args.environmentSuffix}`,
          ...defaultTags,
        },
      },
      { parent: this }
    );

    // Create route to Internet Gateway
    new aws.ec2.Route(
      `public-route-${args.environmentName}-${args.environmentSuffix}`,
      {
        routeTableId: publicRouteTable.id,
        destinationCidrBlock: '0.0.0.0/0',
        gatewayId: this.internetGateway.id,
      },
      { parent: this }
    );

    // Associate public subnets with public route table
    this.publicSubnets.forEach((subnet, index) => {
      new aws.ec2.RouteTableAssociation(
        `public-rta-${args.environmentName}-${index}-${args.environmentSuffix}`,
        {
          subnetId: subnet.id,
          routeTableId: publicRouteTable.id,
        },
        { parent: this }
      );
    });

    // Create Private Route Tables (one per AZ)
    args.availabilityZones.forEach((az, index) => {
      const privateRouteTable = new aws.ec2.RouteTable(
        `private-rt-${args.environmentName}-${az}-${args.environmentSuffix}`,
        {
          vpcId: this.vpc.id,
          tags: {
            Name: `${args.environmentName}-private-rt-${az}-${args.environmentSuffix}`,
            ...defaultTags,
          },
        },
        { parent: this }
      );

      new aws.ec2.Route(
        `private-route-${args.environmentName}-${az}-${args.environmentSuffix}`,
        {
          routeTableId: privateRouteTable.id,
          destinationCidrBlock: '0.0.0.0/0',
          natGatewayId: this.natGateways[index].id,
        },
        { parent: this }
      );

      new aws.ec2.RouteTableAssociation(
        `private-rta-${args.environmentName}-${az}-${args.environmentSuffix}`,
        {
          subnetId: this.privateSubnets[index].id,
          routeTableId: privateRouteTable.id,
        },
        { parent: this }
      );
    });

    // Create Web Tier Security Group
    this.webSecurityGroup = new aws.ec2.SecurityGroup(
      `web-sg-${args.environmentName}-${args.environmentSuffix}`,
      {
        vpcId: this.vpc.id,
        description: 'Security group for web tier allowing HTTPS traffic',
        ingress: [
          {
            protocol: 'tcp',
            fromPort: 443,
            toPort: 443,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow HTTPS from anywhere',
          },
        ],
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow all outbound traffic',
          },
        ],
        tags: {
          Name: `${args.environmentName}-web-sg-${args.environmentSuffix}`,
          ...defaultTags,
        },
      },
      { parent: this }
    );

    // Create App Tier Security Group
    this.appSecurityGroup = new aws.ec2.SecurityGroup(
      `app-sg-${args.environmentName}-${args.environmentSuffix}`,
      {
        vpcId: this.vpc.id,
        description:
          'Security group for app tier allowing traffic from web tier',
        tags: {
          Name: `${args.environmentName}-app-sg-${args.environmentSuffix}`,
          ...defaultTags,
        },
      },
      { parent: this }
    );

    // Add ingress rule to app security group referencing web security group
    new aws.ec2.SecurityGroupRule(
      `app-sg-rule-${args.environmentName}-${args.environmentSuffix}`,
      {
        type: 'ingress',
        fromPort: 0,
        toPort: 65535,
        protocol: 'tcp',
        sourceSecurityGroupId: this.webSecurityGroup.id,
        securityGroupId: this.appSecurityGroup.id,
        description: 'Allow all TCP traffic from web tier',
      },
      { parent: this }
    );

    // Add egress rule to app security group
    new aws.ec2.SecurityGroupRule(
      `app-sg-egress-${args.environmentName}-${args.environmentSuffix}`,
      {
        type: 'egress',
        fromPort: 0,
        toPort: 0,
        protocol: '-1',
        cidrBlocks: ['0.0.0.0/0'],
        securityGroupId: this.appSecurityGroup.id,
        description: 'Allow all outbound traffic',
      },
      { parent: this }
    );

    // Create IAM Role for VPC Flow Logs
    const flowLogsRole = new aws.iam.Role(
      `flow-logs-role-${args.environmentName}-${args.environmentSuffix}`,
      {
        assumeRolePolicy: JSON.stringify({
          Version: '2012-10-17',
          Statement: [
            {
              Action: 'sts:AssumeRole',
              Principal: {
                Service: 'vpc-flow-logs.amazonaws.com',
              },
              Effect: 'Allow',
            },
          ],
        }),
        tags: {
          Name: `${args.environmentName}-flow-logs-role-${args.environmentSuffix}`,
          ...defaultTags,
        },
      },
      { parent: this }
    );

    // Create inline policy for Flow Logs role instead of using managed policy
    new aws.iam.RolePolicy(
      `flow-logs-policy-${args.environmentName}-${args.environmentSuffix}`,
      {
        role: flowLogsRole.id,
        policy: JSON.stringify({
          Version: '2012-10-17',
          Statement: [
            {
              Effect: 'Allow',
              Action: [
                'logs:CreateLogGroup',
                'logs:CreateLogStream',
                'logs:PutLogEvents',
                'logs:DescribeLogGroups',
                'logs:DescribeLogStreams',
              ],
              Resource: '*',
            },
          ],
        }),
      },
      { parent: this }
    );

    // Create CloudWatch Log Group
    this.flowLogGroup = new aws.cloudwatch.LogGroup(
      `flow-logs-${args.environmentName}-${args.environmentSuffix}`,
      {
        name: `/aws/vpc/flow-logs-${args.environmentName}-${args.environmentSuffix}`,
        retentionInDays: 7,
        tags: {
          Name: `${args.environmentName}-flow-logs-${args.environmentSuffix}`,
          ...defaultTags,
        },
      },
      { parent: this }
    );

    // Create VPC Flow Logs
    new aws.ec2.FlowLog(
      `vpc-flow-log-${args.environmentName}-${args.environmentSuffix}`,
      {
        vpcId: this.vpc.id,
        trafficType: 'ALL',
        logDestinationType: 'cloud-watch-logs',
        logDestination: this.flowLogGroup.arn,
        iamRoleArn: flowLogsRole.arn,
        tags: {
          Name: `${args.environmentName}-flow-log-${args.environmentSuffix}`,
          ...defaultTags,
        },
      },
      { parent: this }
    );

    this.registerOutputs({
      vpcId: this.vpc.id,
      publicSubnetIds: this.publicSubnets.map(s => s.id),
      privateSubnetIds: this.privateSubnets.map(s => s.id),
      webSecurityGroupId: this.webSecurityGroup.id,
      appSecurityGroupId: this.appSecurityGroup.id,
      flowLogGroupName: this.flowLogGroup.name,
    });
  }
}
